src/elements/element.rectangle.js

```diff
@@ -78,71 +80,102 @@ module.exports = Element.extend({
-			// borderWidth shold be less than bar width and bar height.
-			var barSize = Math.min(Math.abs(left - right), Math.abs(top - bottom));
-			borderWidth = borderWidth > barSize ? barSize : borderWidth;
-			var halfStroke = borderWidth / 2;
-			// Adjust borderWidth when bar top position is near vm.base(zero).
-			var borderLeft = left + (borderSkipped !== 'left' ? halfStroke * signX : 0);
-			var borderRight = right + (borderSkipped !== 'right' ? -halfStroke * signX : 0);
-			var borderTop = top + (borderSkipped !== 'top' ? halfStroke * signY : 0);
-			var borderBottom = bottom + (borderSkipped !== 'bottom' ? -halfStroke * signY : 0);
-			// not become a vertical line?
-			if (borderLeft !== borderRight) {
-				top = borderTop;
-				bottom = borderBottom;
-			}
-			// not become a horizontal line?
-			if (borderTop !== borderBottom) {
-				left = borderLeft;
-				right = borderRight;
-			}
-		}
-
-		ctx.beginPath();
 		ctx.fillStyle = vm.backgroundColor;
 		ctx.strokeStyle = vm.borderColor;
-		ctx.lineWidth = borderWidth;
-
-		// Corner points, from bottom-left to bottom-right clockwise
-		// | 1 2 |
-		// | 0 3 |
-		var corners = [
-			[left, bottom],
-			[left, top],
-			[right, top],
-			[right, bottom]
-		];
-
-		// Find first (starting) corner with fallback to 'bottom'
-		var borders = ['bottom', 'left', 'top', 'right'];
-		var startCorner = borders.indexOf(borderSkipped, 0);
-		if (startCorner === -1) {
-			startCorner = 0;
-		}
-
-		function cornerAt(index) {
-			return corners[(startCorner + index) % 4];
-		}
 
-		// Draw rectangle from 'startCorner'
-		var corner = cornerAt(0);
-		ctx.moveTo(corner[0], corner[1]);
-
-		for (var i = 1; i < 4; i++) {
-			corner = cornerAt(i);
-			ctx.lineTo(corner[0], corner[1]);
-		}
-
-		ctx.fill();
 		if (borderWidth) {
-			ctx.stroke();
+			if (typeof borderSkipped === 'string') {
+				borderSkipped = {};
+				borderSkipped[vm.borderSkipped] = true;
+			}
+
+			var maxWidth = Math.abs(left - right);
+			var maxHeight = Math.abs(top - bottom);
+			if (!helpers.isObject(borderWidth)) {
+				borderWidth = {
+					bottom: borderSkipped.bottom ? 0 : Math.min(borderWidth, maxHeight),
+					left: borderSkipped.left ? 0 : Math.min(borderWidth, maxWidth),
+					top: borderSkipped.top ? 0 : Math.min(borderWidth, maxHeight),
+					right: borderSkipped.right ? 0 : Math.min(borderWidth, maxWidth)
+				};
+			}
+			ctx.fillRect(
+				left + borderWidth.left * signX,
+				bottom - borderWidth.bottom * signY,
+				right - left - signX * (borderWidth.left + borderWidth.right),
+				top - bottom + signY * (borderWidth.top + borderWidth.bottom));
+
+			ctx.beginPath();
+			if (borderWidth.bottom) {
+				y = bottom - signY * borderWidth.bottom / 2;
+				ctx.moveTo(right, y);
+				if (borderWidth.left) {
+					ctx.lineTo(left + signX * borderWidth.left / 2, y);
+				} else {
+					ctx.lineTo(left, y);
+				}
+				if (borderWidth.bottom !== (borderWidth.left || borderWidth.top || borderWidth.right)) {
```

Comment:
I think we only need to `ctx.stroke()` if `borderWidth.bottom !== borderWidth.left` as a slight optimization. We need to `beginPath` if any of the remaining borders need to be drawn though`